Fix VeSync air_quality fan attribute

jetpacktuxedo opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
A recent change to pyvesync (introduced in 2.02) changed `air_quality` to refer to air quality as an integer representation of perceived air quality rather than a direct reading of the PM2.5 sensor. This was reported as #71427. With an upgrade to [`pyvesync==2.03`](https://github.com/webdjoe/pyvesync/releases/tag/2.03) the PM2.5 sensor access was restored as `air_quality_value`. Unfortunately, `air_quality_value` was not added as an attribute on the fan object, and rather only exists in the `details` dictionary on the fan object.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [x] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #71427
- This PR is related to issue: #71427
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

jetpacktuxedo committed:
Refactor attribute inclusion for VeSync fans.

A recent change to pyvesync (introduced in 2.2) changed `air_quality` to
refer to air quality as an integer representation of perceived air
quality rather than a direct reading of the PM2.5 sensor. With 2.3 the
PM2.5 sensor access was restored as `air_quality_value`. Unfortunately,
`air_quality_value` was not added as an attribute on the fan object, and
rather only exists in the `details` dictionary on the fan object.

homeassistant commented:
Hi @jetpacktuxedo, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#71771).

Once you do that we will be able to review and accept this pull request.

Thanks!

jetpacktuxedo commented:
I haven't updated the documentation yet because I am hoping to get confirmation from a `pyvesync` maintainer that the approach I took here is reasonable to them :)

MartinHjelmare commented on homeassistant/components/vesync/fan.py:
```diff
@@ -157,30 +157,7 @@ def unique_info(self):
-            attr["filter_life"] = self.smartfan.filter_life
-
-        return attr
+        return self.smartfan.details
```
We need to set the keys of the state attributes explicitly so we have control of them.

We're also moving away from using state attributes for all measurements that are relevant on their own. All of these attributes except probably `mode` should preferably be separate sensor entities instead.

jetpacktuxedo replied:
> We need to set the keys of the state attributes explicitly so we have control of them.

Ahh, that's reasonable. I'll switch it back to being explicit.

> We're also moving away from using state attributes for all measurements that are relevant on their own. All of these attributes except probably mode should preferably be separate sensor entities instead.

This is great to hear from a user perspective, jamming stuff into `attributes` is kind of annoying (I had to make a sensor locally to track the `air_quality` attribute over time). That's going to make this a much bigger change though 😭. I'll see what I can do.

Thanks for the quick feedback!

MartinHjelmare commented on homeassistant/components/vesync/fan.py:
```diff
@@ -174,6 +174,9 @@ def extra_state_attributes(self):
             attr["air_quality"] = self.smartfan.air_quality
 
+        if self.smartfan.details.get("air_quality_value") is not None:
+            attr["air_quality_value"] = self.smartfan.details["air_quality"]
```
```suggestion
            attr["air_quality_value"] = self.smartfan.details["air_quality_value"]
```

jetpacktuxedo replied:
I was actually up-in-the-air about this as well. I considered renaming the value that was added recently `air_quality_simple` and making `air_quality` the pm2.5 value again (like it was in previous releases), but that seemed like it might be confusing for people working on `pyvesync`.

I also considered naming the pm2.5 value `pm25` (or similar) in order to make it more clear what it actually represented. Definitely open to suggestions there, though.

jetpacktuxedo replied:
Maybe `particulate_matter_2_5` to match the entity that it will likely eventually correspond to?

jetpacktuxedo replied:
For what it's worth, I think the new `air_quality` metric has value here as well, because I think a few of the supported fans have some sort of vague air quality measurement that isn't pm2.5-based and doesn't report raw sensor value (specifically, the `LV-PUR131S` seems to report with strings such as `Excellent`). I think that is why it was added/changed in `pyvesync` to begin with, but I only have a 400s (pm2.5 sensor) and a 200s (no air quality measurement at all), so I can't really confirm that that is the case or whether it is useful and am only working off of assumptions from reading the `pyvesync` source and documentation.

jetpacktuxedo replied:
I'm more than happy to swap this again, but I did a quick revision using `pm25` as the attribute name and deployed that change to my home instance to make sure everything was working as expected. Here are my attributes locally with this patch in place:
![image](https://user-images.githubusercontent.com/3813932/168320082-c0aa1369-c4a4-4b86-872c-4467d5938aaa.png)

MartinHjelmare replied:
I think we should just restore the previous behavior since we want to remove these attributes in the future and expose the measurements as individual sensors instead.

jetpacktuxedo replied:
Sounds fair enough to me, coming right up 👍

jetpacktuxedo replied:
Alright, from the perspective of a Home Assistant user everything is now back to exactly the way it was before :)
![image](https://user-images.githubusercontent.com/3813932/168325015-979cb210-220a-4af6-a452-4c95f2303c92.png)

jetpacktuxedo committed:
Update homeassistant/components/vesync/fan.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

jetpacktuxedo committed:
Rename `air_quality_value` attribute to `pm25`

This should make it more clear what the attribute actually represents

jetpacktuxedo committed:
`air_quality` attribute reports `air_quality_value`

This restores previous behavior for this integration to what it was
before the `pyvesync==2.02` upgrade, using the `air_quality` attribute
to report pm2.5 concentrations (formerly `air_quality`) rather the
vague measurement now reported by `air_quality`.

MartinHjelmare reviewed (approved):
Thanks!
